src/transformers/trainer.py

```diff
@@ -1142,6 +1152,24 @@ def get_optimizer_cls_and_kwargs(args: TrainingArguments) -> Tuple[Any, Any]:
                 optimizer_kwargs.update(adam_kwargs)
             except ImportError:
                 raise ValueError("Trainer tried to instantiate bnb Adam8bit but bnb is not installed!")
+        elif args.optim == OptimizerNames.ANYPRECISION_ADAMW:
+            try:
+                from torchdistx.optimizers.anyprecision_optimizer import AnyPrecisionAdamW
+
+                optimizer_cls = AnyPrecisionAdamW
+                optimizer_kwargs.update(adam_kwargs)
+                optimizer_kwargs.update(
+                    {
+                        "use_kahan_summation": strtobool(optim_args.get("use_kahan_summation", "False")),
+                        "momentum_dtype": getattr(torch, optim_args.get("momentum_dtype", "float32")),
+                        "variance_dtype": getattr(torch, optim_args.get("variance_dtype", "bfloat16")),
+                        "compensation_buffer_dtype": getattr(
+                            torch, optim_args.get("compensation_buffer_dtype", "bfloat16")
+                        ),
```

Comment:
Let's discuss which defaults would be the most beneficial so that it shines out of the box.

Probably `use_kahan_summation=True`, no?

Reply:
You're right, I just made them as the default in `AnyPrecisionOptimizer`.

Reply:
Hi @stas00  and @atturaioe,
The best/most impressive results so far are definitely running in pure BF16 (so momentum and var set to torch.bfloat16) and use_kahan_summation=True. 

The caveat here is that the model itself needs to be independently set outside of the optimizer to BFloat16 to match (i.e. model.to(torch.bfloat16). 
Are you able to tie the model running in bf16 directly so the user does not have to do that part?

Reply:
unfortunately not for training, but yes for eval, here are the different available  automatic dtype knobs -

https://github.com/huggingface/transformers/blob/4157e3cd7e2bb5a7be6dc065a3e20c49cc1300ab/src/transformers/training_args.py#L271-L291

 `--bf16` here means mixed precision.

-----

so practically what will happen to master weights then? In mixed precision you'd want to have master weights in fp32 - if we switch to `model.to(torch.bfloat16)` this can be quite lossy - as there will be no kahan summation here. it's possible I'm missing some important nuance here.

----

Specifically to what we have right now - perhaps we can for now mark this optimizer as experimental and tune it up as users start to use it?

Reply:
Thanks for this info @stas00 
1 - For mixed precision - you could either 
a - run with the current defaults (M=fp32, Var = BF16, Kahan = False) and that would provide the memory and speed improvements from the Variance in BF16.  That works nicely, and you can make that all work 'automatically' per above control options.
b - you could also go all BF16 (M=BF16, Var = BF16, Kahan = False) because you will still get high precision weight updates with the master weights being in fp32.  This is not as well tested though, but is something we are going to enable in FSDP soon by moving the working weight gradients to BF16, meaning you only have FP32 weights, nothing else. 

To your question - having the weights in BF16 (via model.to) will only work if Kahan summation is active.  If you don't run it with Kahan, then you are exactly right, you will hit weight stagnation and it will not be performant.  
The addition of Kahan is what makes it all work nicely.  

Re: mark as experimental and tune as users run with it - that sounds like a great idea.    I would just go ahead and use the current defaults then (M=FP32, Var = BF16, Kahan = False) as it's plug and play into FP32 or BF16 mixed precision.
I'm working on a video tutorial now actually for this optimizer.  Maybe we can add to the video once this PR is in, and show people how to run it with the manual change of model.to() and setting the defaults directly to get people comfortable with running in pure BF16.

Reply:
That's perfect, Less. Let's do what you suggest.

Let me copy your last comments out of this conversation to a normal comment, so that when this gets resolved it won't disappear. as we will want to eventually cover all these use-cases out of the box.

@atturaioe, so let's keep your defaults for now. and then tweak those in the future.